zhash_insert()/zhashx_insert() return EEXIST

chu11 opened the pull request:
Problem: zhash_insert() and zhashx_insert() return -1 when a duplicate entry exists, NOT when the process is out of memory.  However, a number of locations assume it returns when out of memory and set errno = ENOMEM as a result.

Do not assume zhash_insert()/zhashx_insert() fail on memory allocation errors.  Always ssume duplicate entry.  When necessary set/return errno = EEXIST instead of ENOMEM.

Fixes https://github.com/flux-framework/flux-core/issues/5216

Notes:

My initial reason for doing this was just to make things consistent b/c there was inconsistency in `flux-core`.   Don't know how necessary this is ... Maybe this was just a pointless excercise ... but since I did it, here's the PR :P

There were obviously borderline cases.  Maybe another approach would be to set errno = ENOMEM when a `zhash_insert()` is done directly after a `zhash_lookup()`?

garlick commented on src/broker/groups.c:
```diff
@@ -123,7 +123,7 @@ static struct group *group_lookup (struct groups *g,
         if (zhashx_insert (g->groups, group->name, group) < 0) {
             group_destroy (group);
-            errno = ENOMEM;
+            errno = EEXIST;
```
This is inside a block that has already checked whether the group name is in the hash, so I think the insert return code can be ignored.

garlick commented on src/broker/runat.c:
```diff
@@ -433,6 +433,7 @@ static int runat_push (struct runat *r,
         if (!(entry = runat_entry_create (name)))
             return -1;
         if (zhashx_insert (r->entries, name, entry) < 0) {
+            errno = EEXIST;
```
same here, just ignore the error

garlick commented on src/broker/service.c:
```diff
@@ -161,7 +161,7 @@ int service_add (struct service_switch *sh, const char *name,
     svc->cb = cb;
     svc->cb_arg = arg;
     if (zhash_insert (sh->services, name, svc) < 0) {
-        errno = ENOMEM;
+        errno = EEXIST;
```
same here, ignore the error

garlick commented on src/common/librlist/rlist.c:
```diff
@@ -1612,7 +1612,7 @@ static zhashx_t *rlist_properties (struct rlist *rl)
                      */
                     if (zhashx_insert (properties, name, ids) < 0) {
                         idset_destroy (ids);
-                        errno = ENOMEM;
+                        errno = EEXIST;
```
Just ignore return value and delete the comment above that explains the check.

garlick commented on src/common/librouter/disconnect.c:
```diff
@@ -174,6 +174,7 @@ int disconnect_arm (struct disconnect *dcon, const flux_msg_t *msg)
         if (!(dmsg = disconnect_msg (msg)))
             return -1;
         if (zhashx_insert (dcon->hash, key, dmsg) < 0) {
+            errno = EEXIST;
```
Inside block that already checked.  Just ignore return value.

garlick commented on src/modules/job-list/idsync.c:
```diff
@@ -183,8 +183,10 @@ static int idsync_add_waiter (struct idsync_ctx *isctx,
 
-        if (zhashx_insert (isctx->waits, &isd->id, list_isd) < 0)
-            goto enomem;
+        if (zhashx_insert (isctx->waits, &isd->id, list_isd) < 0) {
+            errno = EEXIST;
+            goto cleanup;
```
ignore return value, already checked item is not in hash

garlick commented on src/modules/job-list/stats.c:
```diff
@@ -72,7 +72,7 @@ static struct job_stats *queue_stats_lookup (struct job_stats_ctx *statsctx,
         if (zhashx_insert (statsctx->queue_stats, job->queue, stats) < 0) {
             flux_log (statsctx->h,
                       LOG_ERR,
-                      "%s: zhashx_insert: out of memory",
+                      "%s: zhashx_insert: queue already exists",
                       __FUNCTION__);
```
Can't happen, already checked.  Just ignore return value.

garlick commented on src/modules/job-manager/event.c:
```diff
@@ -1085,13 +1085,7 @@ int event_index (struct event *event, const char *name)
-             *   case here in case assert-on-malloc failure is fixed
-             *   in the future for zhashx.
-             */
-            errno = ENOMEM;
+            errno = EEXIST;
```
already checked, just ignore result

garlick commented on src/modules/job-manager/plugins/limit-duration.c:
```diff
@@ -94,7 +94,7 @@ static double queues_lookup (zhashx_t *queues, const char *name)
 static int queues_insert (zhashx_t *queues, const char *name, double duration)
 {
     if (zhashx_insert (queues, name, &duration) < 0) { // dups duration
-        errno = ENOMEM;
+        errno = EEXIST;
```
fix the caller too - it hardwires "out of memory".

chu11 replied:
ahh good catch on this one and one below.  Having the `queues_insert()` just return void is probably for best.

garlick commented on src/modules/job-manager/plugins/perilog.c:
```diff
@@ -91,6 +91,7 @@ static struct perilog_proc * perilog_proc_create (flux_plugin_t *p,
     proc->prolog = prolog;
     proc->sp = sp;
     if (zhashx_insert (perilog_config.processes, &proc->id, proc) < 0) {
+        errno = EEXIST;
         free (proc);
         return NULL;
     }
```
Set errno after free to avoid it getting clobbered.

garlick commented on src/modules/job-manager/queue.c:
```diff
@@ -458,10 +458,13 @@ static int queue_configure (const flux_conf_t *conf,
-                    jobq_destroy (q);
+                if (!(q = jobq_create (name)))
                     goto nomem;
+                if (zhashx_insert (queue->named, name, q) < 0) {
+                    jobq_destroy (q);
+                    errprintf (error, "error adding new queue");
+                    errno = EEXIST;
+                    return -1;
```
already checked with zhashx_lookup.  Just ignore return value of insert.

garlick commented on src/modules/kvs-watch/kvs-watch.c:
```diff
@@ -928,6 +928,7 @@ struct ns_monitor *namespace_monitor (struct watch_ctx *ctx,
             return NULL;
         if (zhash_insert (ctx->namespaces, ns, nsm) < 0) {
             namespace_destroy (nsm);
+            errno = EEXIST;
             return NULL;
         }
```
already checked, ignore return value.

garlick reviewed (changes requested):
In several cases simply changing the code to `(void)zhashx_insert()` is appropriate since the key is known not to be in the hash.

chu11 commented:
re-pushed per @garlick comments above, mostly updating to ignore return from `zhash_insert()` after `zhash_lookup()`.  In some job-manager plugins, the result was making a `queue_insert()` function return.  I found one other case of `zhash_insert()` after a `zhash_lookup()` that I corrected in `flux-exec`.  

I also noticed a dumb issue in `cmd/flux-exec` at the same time and fixed it with a commit on top.

garlick commented on src/cmd/flux-exec.c:
```diff
@@ -118,7 +118,7 @@ void completion_cb (flux_subprocess_t *p)
                 log_err_exit ("idset_create");
             (void)zhashx_insert (exitsets, buf, idset);
             if (!zhashx_freefn (exitsets, buf, idset_destroy_wrapper))
-                log_err_exit ("zhashx_freefn");
+                log_msg_exit ("zhashx_freefn");
         }
```
Seems like this return value should be ignored, since zhashx_freefn() only returns NULL when the item doesn't exist, but we just inserted it so we know it does.

chu11 replied:
sounds good, tweaked that last commit to ignore the return value instead.

garlick commented:
Sorry I guess I didn't get back to this for a final approval.  Want to rebase and fix that conflict and then I'll give it a quick final pass?

chu11 commented:
@garlick doh!  i had missed that there was a conflict given a recent merge, thanks for the heads up.  rebased and re-pushed.  I just realized I re-pushed before doing a run of the testsuite locally.  Hopefully I did not re-push too aggressively :P

garlick reviewed (approved):
LGTM! Thanks!
